File: src/practice_robotarm_cpp_examples/src/simple_service_client.cpp
#include <rclcpp/rclcpp.hpp>
#include <memory>
#include <practice_robotarm_msgs/srv/add_two_ints.hpp>

using namespace std::chrono_literals;
using std::placeholders::_1;

class SimpleServiceClient : public rclcpp::Node
{
    public:
        SimpleServiceClient(int a, int b) : Node("simple_service_client")
        {
            client_ = create_client<practice_robotarm_msgs::srv::AddTwoInts>("add_two_ints");
            auto request = std::make_shared<practice_robotarm_msgs::srv::AddTwoInts::Request>();
            request->a = a;
            request->b = b;

            while(!client_->wait_for_service(1s))
            {
                RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Service not available, waiting...");
                if(!rclcpp::ok())
                {
                    RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Interrupted while waiting for service. Exiting.");
                    return;
                }
            }

            auto result = client_->async_send_request(request, std::bind(&SimpleServiceClient::responseCallback, this, std::placeholders::_1));

        }
    
    private:
        rclcpp::Client<practice_robotarm_msgs::srv::AddTwoInts>::SharedPtr client_;

        void responseCallback(rclcpp::Client<practice_robotarm_msgs::srv::AddTwoInts>::SharedFuture future)
        {
            if(future.valid())
            {
                RCLCPP_INFO_STREAM(rclcpp::get_logger("rclcpp"), "Server Response: " << future.get()->sum);
            }
            else
            {
                RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Service Failed to respond.");
            }
        }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    if(argc != 3)
    {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "Usage: simple_service_client <a> <b>");
        return 1;
    }

    auto node = std::make_shared<SimpleServiceClient>(atoi(argv[1]), atoi(argv[2]));
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/practice_robotarm_py_examples/practice_robotarm_py_examples/simple_service_client.py
import rclpy
import sys
from rclpy.node import Node
from practice_robotarm_msgs.srv import AddTwoInts

class SimpleServiceClient(Node):

    def __init__(self, a, b):
        super().__init__('simple_service_client')

        self.client_ = self.create_client(AddTwoInts, 'add_two_ints')

        while not self.client_.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('service not available, waiting more...')
        
        self.req_ = AddTwoInts.Request()
        self.req_.a = a
        self.req_.b = b
        self.future_ = self.client_.call_async(self.req_)
        self.future_.add_done_callback(self.responseCallback)

    def responseCallback(self, future):
        self.get_logger().info('Service Response: %d' % future.result().sum)

def main(args=None):
    rclpy.init()
    if len(sys.argv) != 3:
        print('usage: simple_service_client.py a b')
        return -1

    simple_service_client = SimpleServiceClient(int(sys.argv[1]), int(sys.argv[2]))
    rclpy.spin(simple_service_client)
    simple_service_client.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()